include/dali/core/tensor_layout.h

```diff
@@ -80,11 +75,20 @@ class TensorLayout {
   TensorLayout(const std::string &s) : TensorLayout(s.data(), s.length()) {  // NOLINT
   }
 
+  /** @brief Returns a reference to d-th dimension in the layout
+   * @remarks Any of the following leads to undefined behavior:
+   *    - accessing values at d > ndim()
+   *    - changing '\0' terminator to any other value
+   *    - replacing any meaningful character with '\0'
+   */
   DALI_HOST_DEV
   constexpr char &operator[](int d) noexcept {
     return data_[d];
   }
 
+  /** @brief Returns a reference to d-th dimension in the layout
+   * @remarks Value at ndim() is always '\0', accessing values beyond ndim() is forbidden.
+   */
```

Comment:
I'm nitpicking but you have two different remarks for operator[] depending on the constness.